### PR TITLE
Fixes for pbench-list-tools

### LINKED ISSUE #2302: pbench-list-tools (v0.71 - python version) produces a traceback  for an unknown group
It should probably catch the exception and exit with an error message. Instead it does this:

```
# pbench-list-tools --group=bar
Traceback (most recent call last):
  File "/opt/pbench-agent/lib/pbench/agent/tool_group.py", line 43, in verify_tool_group
    tg_dir = tg_dir_name.resolve(strict=True)
  File "/usr/lib64/python3.9/pathlib.py", line 1204, in resolve
    s = self._flavour.resolve(self, strict=strict)
  File "/usr/lib64/python3.9/pathlib.py", line 362, in resolve
    return _resolve(base, str(path)) or sep
  File "/usr/lib64/python3.9/pathlib.py", line 346, in _resolve
    target = accessor.readlink(newpath)
  File "/usr/lib64/python3.9/pathlib.py", line 451, in readlink
    return os.readlink(path)
FileNotFoundError: [Errno 2] No such file or directory: '/var/lib/pbench-agent/tools-v1-bar'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/pbench-agent/util-scripts/pbench-list-tools", line 10, in <module>
    sys.exit(main())
  File "/usr/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python3.9/site-packages/click/decorators.py", line 73, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/usr/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/opt/pbench-agent/lib/pbench/cli/agent/commands/tools/list.py", line 135, in main
    status = ListTools(ctxt).execute()
  File "/opt/pbench-agent/lib/pbench/cli/agent/commands/tools/list.py", line 39, in execute
    for path in self.gen_tools_group_dir(group).glob("*/**"):
  File "/opt/pbench-agent/lib/pbench/agent/base.py", line 124, in gen_tools_group_dir
    return ToolGroup.verify_tool_group(group, pbench_run=self.pbench_run)
  File "/opt/pbench-agent/lib/pbench/agent/tool_group.py", line 45, in verify_tool_group
    raise BadToolGroup(
pbench.agent.tool_group.BadToolGroup: Bad tool group, 'bar': directory /var/lib/pbench-agent/tools-v1-bar does not exist
```

### LINKED ISSUE #2345: pbench-list-tools ends with exception
pbench-agent-0.71.0-185g5527f23f1.noarch

1) run
```
pbench-register-tool-trigger --start-trigger='START TRIGGER' --stop-trigger='STOP TRIGGER' --group=another 
```
2) run
```
pbench-list-tools -n perf
```
```
Traceback (most recent call last):
  File "/opt/pbench-agent/util-scripts/pbench-list-tools", line 10, in <module>
    sys.exit(main())
  File "/usr/lib/python3.6/site-packages/click/core.py", line 721, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 894, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args[1:], **kwargs)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/opt/pbench-agent/lib/pbench/cli/agent/commands/tools/list.py", line 135, in main
    status = ListTools(ctxt).execute()
  File "/opt/pbench-agent/lib/pbench/cli/agent/commands/tools/list.py", line 61, in execute
    if self.context.name in self.tools(path):
  File "/opt/pbench-agent/lib/pbench/cli/agent/commands/tools/base.py", line 19, in tools
    for p in path.iterdir()
  File "/opt/pbench-agent/lib/pbench/cli/agent/commands/tools/base.py", line 18, in <listcomp>
    p.name
  File "/usr/lib64/python3.6/pathlib.py", line 1081, in iterdir
    for name in self._accessor.listdir(self):
  File "/usr/lib64/python3.6/pathlib.py", line 387, in wrapped
    return strfunc(str(pathobj), *args)
NotADirectoryError: [Errno 20] Not a directory: '/var/lib/pbench-agent/tools-v1-another/__trigger__'
```

### LINKED ISSUE #2346: pbench-list-tools with options ends with exception
pbench-agent-0.71.0-185g5527f23f1.noarch

run pbench-list-tools -o

Traceback (most recent call last):
  File "/opt/pbench-agent/util-scripts/pbench-list-tools", line 10, in <module>
    sys.exit(main())
  File "/usr/lib/python3.6/site-packages/click/core.py", line 721, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 894, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args[1:], **kwargs)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/opt/pbench-agent/lib/pbench/cli/agent/commands/tools/list.py", line 135, in main
    status = ListTools(ctxt).execute()
  File "/opt/pbench-agent/lib/pbench/cli/agent/commands/tools/list.py", line 42, in execute
    p.read_text() for p in self.tools(path)
  File "/opt/pbench-agent/lib/pbench/cli/agent/commands/tools/list.py", line 42, in <listcomp>
    p.read_text() for p in self.tools(path)
AttributeError: 'str' object has no attribute 'read_text'

### PR DESCRIPTION
**Pull request by ndokos**
Fixes #2302
Fixes #2345 
Fixes #2346 

This PR consists of five commits:

- the first commit adds some basic functional tests that trigger the uncaught exceptions in the three issues above.
- the next three commits make the changes necessary to fix each one of the issues.

This PR will be followed by a subsequent PR that will fix additional issues in `pbench-list-tools` that do not cause exceptions. No attempt has been made to make the exit status or the output of the program correct: that is left to the subsequent PR. The exit status of this program is, to the extent possible, compatible with the 0.69.9 version.

This PR and the subsequent one replace PR #2453.

### CODE REVIEW

**Comment by webbnh** [comment]
> the fifth commit is just black reformatting.

Since Black's changes are all applied to new code from commit `fac297f`, it would have been better to squash the two commits before pushing the branch for the PR.

Also, this PR apparently needs to be rebased.  (I wouldn't be opposed to your doing the above squash during that operation.)

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py [hunk 1]
```diff
@@ -1,31 +1,106 @@
...
+        assert empty == err
+        assert usage in out
+        assert exitcode == 0
+
+    def test_no_args(self, agent_config):
+        command = ["pbench-list-tools"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert traceback not in err
+        assert empty == err
+        assert b"" == out
```
Comment (webbnh) [inline lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:24]: It seems weird that you would use a literal in line 24 and a variable containing the same value in line 23.

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py [hunk 2]
```diff
@@ -1,31 +1,106 @@
...
-    assert b"default: testhost.example.com ['perf']" in out
-
-    command = ["pbench-list-tools", "-n", "perf"]
-    out, err, exitcode = pytest.helpers.capture(command)
-    assert exitcode == 0
-    assert b"tool name: perf groups: default" in out
+empty = b""
+usage = b"Usage: pbench-list-tools [OPTIONS]"
+bad_group = b"Bad tool group: "
+traceback = b"Traceback (most recent call last):\n"
```
Comment (webbnh) [inline lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:7]: Symbols for values which are intended to be "constant" should have solid-caps names.

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py [hunk 3]
```diff
@@ -1,31 +1,106 @@
...
+
+
+class Test_list_tools_no_tools_registered:
+    def test_help(self):
+        command = ["pbench-list-tools", "--help"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert traceback not in err
+        assert empty == err
```
Comment (webbnh) [inline lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:15]: It seems like the second assertion (at line 15) implies the first (at line 14).  Ditto in most of the other test scenarios, as well.  (I think line 80 is the only case where it's worth checking.)

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py [hunk 4]
```diff
@@ -1,31 +1,106 @@
...
+        assert exitcode == 0
+
+
+class Test_list_tools_tools_registered:
+    @pytest.fixture
+    def tool(self, pbench_run):
+        p = pbench_run / "tools-v1-default" / "testhost.example.com"
+        p.mkdir(parents=True)
+        tool = p / "perf"
+        tool.touch()
```
Comment (webbnh) [inline lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:34]: Are you sure that we want to be doing this "manually", instead invoking `pbench-register-tool`?  (Ditto for `tool_with_random_file()`.)

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py [hunk 5]
```diff
@@ -1,31 +1,106 @@
...
+        assert empty == err
+        assert b"default: testhost.example.com ['perf']" in out
+        assert exitcode == 0
+
+    def test_name(self, tool, agent_config):
+        command = ["pbench-list-tools", "-n", "perf"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert traceback not in err
+        assert empty == err
+        assert b"tool name: perf groups: default" in out
```
Comment (webbnh) [inline lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:65]: This text looks funny without any punctuation (and only a single space) between the tool name (`perf`) and "groups".

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py [hunk 6]
```diff
@@ -1,31 +1,106 @@
...
+        assert exitcode == 0
+
+    # Issue #2345
+    def test_name_with_random_file(self, tool_with_random_file, agent_config):
```
Comment (webbnh) [inline lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:69]: Should we also have a test scenario with a random _directory_ file?

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py [hunk 7]
```diff
@@ -1,31 +1,106 @@
...
+        tool.write_text("--interval=300")
+
+    # Issue #2346
+    def test_existing_group_options(self, single_group_tools, agent_config):
+        command = ["pbench-list-tools", "--with-option"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert traceback not in err
+        assert empty == err and exitcode == 0
+        assert (
+            b"default: testhost.example.com [('iostat', '--interval=30'), ('mpstat', '--interval=300')]\n"
```
Comment (webbnh) [inline lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:104]: I'm looking forward to having more human readable output....

File: lib/pbench/cli/agent/commands/tools/list.py [hunk 8]
```diff
@@ -36,13 +37,18 @@ def execute(self):
...
+                            ]
+                        else:
+                            host_tools[group][path.name] = [p for p in self.tools(path)]
+                except BadToolGroup:
+                    self.logger.error("Bad tool group: %s", group)
+                    # 0.69.9 behavior
+                    return 0
             if host_tools:
                 for k, v in host_tools.items():
                     for h, t in v.items():
```
Comment (webbnh) [inline lib/pbench/cli/agent/commands/tools/list.py:54]: I think we want these traversals in sorted order, for test reproducibility.  (Ditto line 73.)

**Review by webbnh — CHANGES_REQUESTED** [review]
Looks like a good (re-)start.  I have comments on small things.

File: lib/pbench/cli/agent/commands/tools/list.py (hunk 8 above)
Comment (ndokos) [reply lib/pbench/cli/agent/commands/tools/list.py:54]: That will be done in the subsequent PR. All I'm interested in in this PR is getting rid of the tracebacks. I don't really care about the output (yet).

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py (hunk 7 above)
Comment (ndokos) [reply lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:104]: Coming in the subsequent PR - stay tuned.

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py (hunk 6 above)
Comment (ndokos) [reply lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:69]: We probably should: I really have no idea what would happen. I'll add some test cases in the subsequent PR.

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py (hunk 5 above)
Comment (ndokos) [reply lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:65]: I specifically am not addressing output formatting in this PR. That's for the subsequent PR.

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py (hunk 4 above)
Comment (ndokos) [reply lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:34]: Perhaps, although I also want to manufacture situations that `pbench-register-tool` and company cannot create. Ideally, the ToolGroup abstraction would be used (but it is read-only ATM) to manufacture the "legal" things and it's only the hinky ones that would have to be done by hand.

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py (hunk 3 above)
Comment (ndokos) [reply lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:15]: You are right, but I had a tough time seeing the tracebacks in the output for some reason, so this makes sure that the assertion *tells* me it was a traceback.

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py (hunk 1 above)
Comment (ndokos) [reply lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:24]: Consistency was never my strong suit  :-)  will fix.

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py (hunk 2 above)
Comment (ndokos) [reply lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:7]: OK

**Comment by ndokos** [comment]
The black commit has been squashed with the first commit and the branch has been rebased. The fixes to the tests have been made as requested. Fixes to the output or exit status have *NOT* been made: they will be part of a subsequent PR. Some additional tests that were discusses in the comments will also be part of that PR.

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py [hunk 9]
```diff
@@ -1,31 +1,106 @@
...
+        p = pbench_run / "tools-v1-default" / "testhost.example.com"
+        p.mkdir(parents=True)
+        tool = p / "perf"
+        tool.touch()
+
+    @pytest.fixture
+    def tool_with_random_file(self, pbench_run):
+        p = pbench_run / "tools-v1-default" / "testhost.example.com"
+        p.mkdir(parents=True)
+        (p.parent / "foo").touch()
```
Comment (dbutenhof) [inline lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:40]: Should we be testing "random files" at other directory levels, too? E.g., trigger/label file skipping?

**Review by dbutenhof — DISMISSED** [review]
Just a few minor comments on the testing infrastructure here, but I think it looks OK.

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py [hunk 10]
```diff
@@ -1,31 +1,106 @@
...
+        tool = p / "perf"
+        tool.touch()
```
[the remark this answers is not in the document]
Comment (ndokos) [reply lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:42]: Not sure yet, but probably - see your comment below.

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py (hunk 9 above)
Comment (ndokos) [reply lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:40]: Yes, I think so: although most of the test cases arose either out of know bugs or of testing "normal" functionality, there are still a lot of corner cases to test. Actually, I think that the current implementation of `pbench-list-tools` is going to be replaced: the only thing that will be left will be the test suite, so the more complete that is, the better off we'll be. I imagine the test suite  will grow and, in some sense, be a model of what the test suites for other tools will look like.

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (webbnh) [reply lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:42]: I could go either way on this.  Having this setup encapsulated is a feature, especially since this fixture closely follows the one above it.  On the other hand, there is a cost to forcing the reader to go somewhere else to find everything.  But, I think the encapsulation is probably worth the cost.

**Review by webbnh — DISMISSED** [review]
Good by me.

**Review by portante — DISMISSED** [review]

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dbutenhof) [reply lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:42]: I don't know about other IDEs (maybe I'm in the slums and nobody else has this issue), but vscode doesn't see the fixture "parameters" as references (not really surprising, since it'd take special pytest awareness to recognize this), which makes tracking the use or origin of fixtures awkward. It's a great feature, if extraordinarily quirky; but overuse just makes the code harder to read. Still, I'm not objecting, just questioning; so I'll resolve this conversation and let it go. 😁

**Comment by ndokos** [comment]
Rebased and squashed.

**Review by dbutenhof — APPROVED** [review]

File: lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (webbnh) [reply lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py:42]: Yeah, PyCharm doesn't like them much, either.

Of course, there's no huge advantage to this being a "fixture" over being a "helper routine"...the difference is that, as a fixture, it's called implicitly, in exchange for the weird presence in the function signature.

So long as it's not using any of the other superpowers of a fixture (like implicit invocations at broader scopes, clean-up actions, et al.), it could just as easily be a simple function call, and perhaps that would be better.  (But, as you say, this conversation has been resolved, so....)

**Review by webbnh — APPROVED** [review]
Github says that nothing's changed, so it must still be good.
